podresources: list: use active pods

ffromani opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
The podresources API List implementation uses the internal data of the  resource managers as source of truth. But it iterates over the full list of pods known by the kubelet. Differently, all the resource managers do their internal sync using the list of active pods, fetched each reconciliation step using a helper function.
This causes the `List` endpoint to return incorrect data, because it has a view of the world different from what the resource managers actually have.

#### Which issue(s) this PR fixes:
Fixes #132020
Fixes #119423
Fixes #131999

#### Special notes for your reviewer:
We need to acknowledge the the [warning in the docstring of GetActivePods](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_pods.go#L194). Arguably, having the endpoint using a different podset than the resource manager causes way more harm than good because of the desync.
And arguably, it's better to fix this issue in just one place instead of having the `List` use a different pod set for unclear reason. For these reasons, while important, I don't think the warning per se invalidated this change.

We need to further acknowledge the `List` endpoint used the full pod list since its inception. So, we will add a way in the `v1` endpoint to preserve the old behavior to zero the chance to break users of the API depending on the old behavior. 

The old `v1alpha1` endpoint will be not modified intentionally.

#### Does this PR introduce a user-facing change?
```release-note
Change the node-local podresources API endpoint to only consider of active pods. Because this fix changes a long-established behavior, users observing a regressions can use the KubeletPodResourcesListUseActivePods feature gate (default on) to restore the old behavior. Please file an issue if you encounter problems and have to use the Feature Gate.
```

ffromani commented:
the PR is reviewable. I'm finishing the e2e tests.

ffromani commented:
/help

ffromani commented:
/test

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-podresources

harche commented:
/assign

ffromani committed:
e2e: node: memory manager: build everywhere, run only on linux

Since the KEP 4885
(https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/4885-windows-cpu-and-memory-affinity/README.md)
memory manager is supported also on windows.

Plus, we want to add podresources e2e tests which configure
the memory manager. Both these facts suggest it's useful to build
the e2e memory manager tests on all OSes, not just on linux;

However, since we are not sure we are ready to run these tests
everywhere, we tag them LinuxOnly to keep preserve most of the
old behavior.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: e2e: podresources: add more e2e tests

add more e2e tests to cover the interaction with
core resource managers (cpu, memory) and to ensure
proper reporting.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-podresources

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-memory-manager

ffromani commented:
/cc @marquiz 
because your team IIRC consumes this API

ffromani commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -33,7 +33,9 @@ message AllocatableResourcesResponse {
 
 // ListPodResourcesRequest is the request made to the PodResourcesLister service
-message ListPodResourcesRequest {}
+message ListPodResourcesRequest {
+  bool active_only = 1;
+}
```
Because how gRPC and protobuf works, I'm confident this change is purely additive and won't change the existing behavior. But it will make somehow hard to consume the new behavior - how a client willing to opt in into the new behavior knows a kubelet supports it and honor its request?

I think we have a simple fix handy which does NOT require an API bump, which IMVHO would be impractical and overkill:
```
diff --git a/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto b/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto
index 55bbf3415ce..2c1cbe2b51f 100644
--- a/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto
+++ b/staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto
@@ -32,14 +32,21 @@ message AllocatableResourcesResponse {
     repeated ContainerMemory memory = 3;
 }
 
+// Filters restricts the result set returned by the API
+message Filters {
+  bool active_only = 1;
+}
+
 // ListPodResourcesRequest is the request made to the PodResourcesLister service
 message ListPodResourcesRequest {
-  bool active_only = 1;
+  Filters filters = 1;
 }
 
 // ListPodResourcesResponse is the response returned by List function
 message ListPodResourcesResponse {
     repeated PodResources pod_resources = 1;
+    Filters supported_filters = 2;
+    Filters applied_filters = 3;
 }
 
 // PodResources contains information about the node resources assigned to a pod

```
this way:
1. the server will always send the `supported_filters` which will tell the client what it can set; again per gRPC/protobuf spec older clients will just happily ignore the extra fields, so no compatibility breakage;
2. the server will mirror in the `applied_filters` whatever the client set in the request (if at all) so the client can know what it was done and have clear feedback.

The cost, besides some extra fields (negligible) is one extra `List` call *just for new clients which want to use the filters*, which seems pretty fair to me everything considered.

ffromani replied:
Correction: is a bad idea to use the same struct in both request and answer, so the `Filter` Message should be duplicated in two (atm) identical `FilterRequest` and `FilterResponse`. Which adds a bit more of extra noise but doesn't change the core of the proposal

marquiz replied:
How probable do we see the case that someone would depend on the (seemingly buggy) behaviour(?) Just ponder is it absolutely out of the question to have a feature-gate or a kubelet config option for this. So that existing clients would start to get the correct data without any changes. WDYT?

ffromani replied:
We have 2 main problems:
1. the API  was not crisply defined from the beginning. This is one key contributing factors to explain why it took me so long to converge on a fix, we were chasing intermittent issues on the numa-aware scheduling side for many many months and just recently everything clicked. See: https://github.com/kubernetes/kubernetes/issues/119423#issuecomment-1653407887
2. after so much time, I believe https://en.wiktionary.org/wiki/Hyrum%27s_law is now on full effect so I'm reluctant to change the default behavior. The nvidia dgcm exporter and multus are known consumers of this API.
However, I'm open to change the default _and_ introduce a feature gate in general, but I'm very wary of the ramifications

I also thought about adding a parallel but different `List` endpoint but it seems more harm than good to have two almost identical APIs.
Likewise, adding a v2 API and changing the default behavior of `List` doesn't seem much better as we potentially split the userbase creating a python2/python3-like scenario.

Hence, the filtering plus better docs (see comment above) are IMO the best way forward. The changes needed to add the filtering to cover the use case are really small and very contained. Even ensuring discoverability and full backward compatibility, we are in the realm of ~50 lines change, which seems very very acceptable. 90% and more of the PR mass is autogenerated code and test code.

ffromani replied:
> > However, I'm open to change the default and introduce a feature gate in general, but I'm very wary of the ramifications
> 
> Surely, this is an option but for now I think it is best to keep the current (arguably wrong behavior) as default and reevaluate at a later stage based on the adoption.

Yes, we are in agreement. New evidence points out the current behavior is pretty certainly buggy and not the desired behavior (https://github.com/kubernetes/kubernetes/issues/119423#issuecomment-2934790530) but still the current API behaves like this since 2020, and it is stable. I think it's best to opt-in the new, correct behavior, plus we need a way to distinguish fixed and faulty kubelets.

> > the filtering plus better docs (see comment above) are IMO the best way forward
> 
> I like this direction. Filtering introduces a more explicit contract between client and server, which is a solid foundation. One suggestion: instead of a boolean `activeOnly` flag, should we consider a more flexible filtering mechanism (e.g., by pod phase)? That could give us more room for future extensibility without needing to change the API again.

I hear you and I understand the reason behind this proposal. I'm however resisting to the (very appealing) idea to implement a generic filtering mechanism. The current filtering mechanism is implemented in order to
1. allow clients opt-in the new behavior dynamically, on a call-by-call basis if needed
2. enable discoverability
3. while preserving backward compatibility

So the implemented filtering is tailor made for the changes which implement the correct behavior - there's one more pending which I'm not sure we should add: pods which don't have any exclusive resource assigned should not even figure out in the output. Trivial to do both in server and in the client side (and existing clients already do this).

So to recap the current filtering is functional to the 3 goals above, making it generic will totally call for a API extension and would require much more conversation and review.
I'm open to further discussion though.

ffromani commented:
/priority important-longterm

not a regression, has been broken (or at very least incorrect/misleading) for years

marquiz commented:
/cc @PiotrProkop

shajmakh reviewed (commented):
Thanks for this and for extending the tests coverage. 
Can you please clarify in the commit message the new flag name to be adjusted and the expected value for it to activate the resources calculation of active pods only?

ffromani commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -55,7 +55,13 @@ func (p *v1PodResourcesServer) List(ctx context.Context, req *podresourcesv1.Lis
 
-	pods := p.podsProvider.GetPods()
+	var pods []*v1.Pod
+	if req.GetActiveOnly() {
+		pods = p.podsProvider.GetActivePods()
+	} else {
+		pods = p.podsProvider.GetPods()
+	}
```
A log would be >= 6 to avoid to spam the logs. I would strongly advocate for the filtering expansion instead: https://github.com/kubernetes/kubernetes/pull/132028#discussion_r2119222326

ffromani commented on pkg/kubelet/apis/podresources/server_v1.go:
```diff
@@ -55,7 +55,13 @@ func (p *v1PodResourcesServer) List(ctx context.Context, req *podresourcesv1.Lis
 	metrics.PodResourcesEndpointRequestsTotalCount.WithLabelValues("v1").Inc()
```
I will amend the code references to point to a tag.

ffromani commented:
/retest-required

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-podresources

swatisehgal commented on staging/src/k8s.io/kubelet/pkg/apis/podresources/v1/api.proto:
```diff
@@ -32,14 +32,26 @@ message AllocatableResourcesResponse {
 }
 
+// ListPodResourcesFilterRequest restricts the result set returned by the List API
+message ListPodResourcesFilterRequest {
```
Just reinterating my previous [comment](https://github.com/kubernetes/kubernetes/pull/132028#discussion_r2123239726) at the corresponding API change about using a more flexible filtering mechanism (e.g., by pod phase) instead of a boolean `activeOnly` flag.

rphillips commented on pkg/kubelet/apis/podresources/server_v1_test.go:
```diff
@@ -112,7 +112,12 @@ func TestListPodResourcesV1(t *testing.T) {
 			memory:           []*podresourcesapi.ContainerMemory{},
 			dynamicResources: []*podresourcesapi.DynamicResource{},
-			expectedResponse: &podresourcesapi.ListPodResourcesResponse{},
+			expectedResponse: &podresourcesapi.ListPodResourcesResponse{
```
We need to have some tests to test the backwards compatibility.

ffromani replied:
I wonder how to do that but in this very case I can perhaps somehow reuse the v1alpha1 client or somehow ship the previous version of the proto/compiled bindings

ffromani commented:
So, the options on the table are:
1. fix the issues with this approach implemented in this PR
2. fix the issue with a different approach: sync the state before each `List` call, like we do for devices. Still pretty much contained, adds lock contention
3. option 1, add a feature gate, without a KEP (benefits unclear: all changes here are opt-in anyway)
4. option 2, add a feature gate, without a KEP
5. file a new KEP. If we file a new KEP I'd rather propose a new endpoint (`ListActive`? `ListAllocatedResources`?) fixing all the known issues of the `List` endpoint. I really do think this KEP should start beta, as there are no unknowns (and no risks, btw, being a new additive and partially overlapping endpoint)

ffromani commented:
> So, the options on the table are:
> 
>     1. fix the issues with this approach implemented in this PR
> 
>     2. fix the issue with a different approach: sync the state before each `List` call, like we do for devices. Still pretty much contained, adds lock contention
> 
>     3. option 1, add a feature gate, without a KEP (benefits unclear: all changes here are opt-in anyway)
> 
>     4. option 2, add a feature gate, without a KEP
> 
>     5. file a new KEP. If we file a new KEP I'd rather propose a new endpoint (`ListActive`? `ListAllocatedResources`?) fixing all the known issues of the `List` endpoint. I really do think this KEP should start beta, as there are no unknowns (and no risks, btw, being a new additive and partially overlapping endpoint)

alternative take: https://github.com/kubernetes/kubernetes/pull/132081 - only fixes #132020 which is sufficient to fix the most serious issues (and unbreaks the numa-aware scheduler plugin). #132081 also has no backward compatibility or user facing concerns, it's purely an internal change.

swatisehgal commented:
> > The change looks well-contained, and I like the direction overall. That said, since it introduces an API change, we should confirm whether this qualifies as a bug fix or if it needs to follow the KEP process. Let’s make sure we’re aligned on the right path forward process-wise.
> 
> Taking a step back, I'm still convinced this qualifies for a bugfix, but because of the backward compatibility concerns and the fact we need to introduce some filtering, I can also see why the KEP process can be required, like we did for other changes which were small and self contained but required user-visible changes. I really do see good reasons for both approaches so I welcome third party opinions.

I'm personally leaning toward treating this as a bugfix as well, especially since the intent is to preserve existing behavior while adding clarity. That said, the introduction of an API change is why I raised the question. It might be sufficient to go through API review without requiring a full KEP, especially since we’ve had similar cases like [#127525](https://github.com/kubernetes/kubernetes/pull/127525) where a feature gate helped preserve old behavior. This change is arguably even safer. Still, I can’t say definitively whether we can skip the KEP, so I think it’s worth confirming explicitly.

haircommander commented:
what if we introduced an on by default beta feature gate but skipped the KEP process like in https://github.com/kubernetes/kubernetes/pull/127525 ? that gives users an out if it breaks them and gives us time to get feedback, but in a way that has less logistical overhead

ffromani commented:
> what if we introduced an on by default beta feature gate but skipped the KEP process like in #127525 ? that gives users an out if it breaks them and gives us time to get feedback, but in a way that has less logistical overhead

thanks Peter. This is also what @marquiz proposed a couple days ago, so it seems we are trending this direction. I can implement this and we can evaluate. This will allow us to remove all the extra messages.

ffromani committed:
podresources: list: use active pods in list

The podresources API List implementation uses the internal data of the
resource managers as source of truth.
Looking at the implementation here:
https://github.com/kubernetes/kubernetes/blob/v1.34.0-alpha.0/pkg/kubelet/apis/podresources/server_v1.go#L60
we take care of syncing the device allocation data before querying the
device manager to return its pod->devices assignment.
This is needed because otherwise the device manager (and all the other
resource managers) would do the cleanup asynchronously, so the `List` call
will return incorrect data.

But we don't do this syncing neither for CPUs or for memory,
so when we report these we will get stale data as the issue #132020 demonstrates.

For CPU manager, we however have the reconcile loop which cleans the stale data periodically.
Turns out this timing interplay was actually the reason the existing issue #119423 seemed fixed
(see: https://github.com/kubernetes/kubernetes/issues/119423#issuecomment-2786891218).
But it's actually timing. If in the reproducer we set the `cpuManagerReconcilePeriod` to a time
very high (>= 5 minutes), then the issue still reproduces against current master branch
(https://github.com/kubernetes/kubernetes/blob/v1.34.0-alpha.0/test/e2e_node/podresources_test.go#L983).

Taking a step back, we can see multiple problems:
1. not syncing the resource managers internal data before to query for
   pod assignment (no removeStaleState calls) but most importantly
2. the List call iterate overs all the pod known to the kubelet. But the
   resource managers do NOT hold resources for non-running pod, so it is
   better, actually it's correct to iterate only over the active pods.
   This will also avoid issue 1 above.

Furthermore, the resource managers all iterate over the active pods
anyway:
`List` is using all the pods known about:
1. https://github.com/kubernetes/kubernetes/blob/v1.34.0-alpha.0/pkg/kubelet/kubelet.go#L3135 goes in
2. https://github.com/kubernetes/kubernetes/blob/v1.34.0-alpha.0/pkg/kubelet/pod/pod_manager.go#L215

But all the resource managers are using the list of active pods:
1. https://github.com/kubernetes/kubernetes/blob/v1.34.0-alpha.0/pkg/kubelet/kubelet.go#L1666 goes in
2. https://github.com/kubernetes/kubernetes/blob/v1.34.0-alpha.0/pkg/kubelet/kubelet_pods.go#L198

So this change will also make the `List` view consistent with the
resource managers view, which is also a promise of the API currently
broken.

We also need to acknowledge the the warning in the docstring of GetActivePods.
Arguably, having the endpoint using a different podset wrt the resource managers with the
related desync causes way more harm than good.
And arguably, it's better to fix this issue in just one place instead of
having the `List` use a different pod set for unclear reason.
For these reasons, while important, I don't think the warning per se
invalidated this change.

We need to further acknowledge the `List` endpoint used the full pod
list since its inception. So, we will add a Feature Gate to disable this
fix and restore the old behavior. We plan to keep this Feature Gate for
quite a long time (at least 4 more releases) considering how stable this
change was. Should a consumer of the API being broken by this change,
we have the option to restore the old behavior and to craft a more
elaborate fix.

The old `v1alpha1` endpoint will be not modified intentionally.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-podresources

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-podresources

marquiz reviewed (approved):
Thanks @ffromani, LGTM from me, fwiw

ffromani commented on pkg/features/kube_features.go:
```diff
@@ -1890,6 +1911,9 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 		{Version: version.MustParse("1.33"), Default: true, PreRelease: featuregate.Beta},
 	},
+	KubeletPodResourcesListUseActivePods: {
+		{Version: version.MustParse("1.34"), Default: true, PreRelease: featuregate.Beta},
```
Thanks for your review!
Yes, we will remove once sufficient time passed without any issue filed: https://github.com/kubernetes/kubernetes/pull/132028/files#diff-71e3b98f9a6bbf5b8421e26a7ba0c079f397cd8d49abacdad943c66a4f44f03dR1017

I will look in the Deprecated suggestion now.

aojea replied:
see as an example https://github.com/kubernetes/kubernetes/pull/125813 but you can git blame in most of the Deprecated features in the pkg/features/kube_features.go as most of them are being used for this model of rolling out bugfixes on established behaviors in a controlled way

ffromani commented:
uhm: 
```
panic: feature "KubeletPodResourcesListUseActivePods" introduced as deprecated must provide a 1.0 entry indicating initial state
goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.Must(...)
	k8s.io/apimachinery/pkg/util/runtime/runtime.go:293
k8s.io/kubernetes/pkg/features.init.0()
	k8s.io/kubernetes/pkg/features/kube_features.go:1885 +0x138
!!! [0610 13:33:30] Call tree:
!!! [0610 13:33:30]  1: hack/make-rules/test-cmd.sh:196 run_kube_apiserver(...)
junit report dir: /logs/artifacts
+++ [0610 13:33:30] Clean up complete 
```

ffromani commented:
/retest-required

ffromani commented:
/hold cancel

all tests look good now

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-podresources

marquiz reviewed (commented):
/lgtm

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd

ffromani commented:
humble ping :)

ffromani commented:
rebased, no changes in behavior.

ffromani commented:
/test pull-kubernetes-e2e-gce

rphillips commented on pkg/features/kube_features.go:
```diff
@@ -1275,6 +1296,11 @@ var defaultVersionedKubernetesFeatureGates = map[featuregate.Feature]featuregate
 	},
 
+	KubeletPodResourcesListUseActivePods: {
+		{Version: version.MustParse("1.0"), Default: true, PreRelease: featuregate.GA},
```
Should this be false, then marked as default true in 1.34?

ffromani replied:
thanks @rphillips , the approach you suggested aligns more with how the deprecated FGs, even the one guarding fixes, seems to be used. Changing as suggested.

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-cos-cgroupv2-containerd-node-e2e-serial

ffromani commented:
ping for review @mrunalp @SergeyKanzhelev

SergeyKanzhelev reviewed (approved):
/lgtm
/approve
